packages/dds/tree/src/util/rangeMap.ts

```diff
@@ -103,67 +166,110 @@ export class RangeMap<T> {
 	 * @param start - The start of the range to delete (inclusive).
 	 * @param length - The length of the range to delete.
 	 */
-	public delete(start: number, length: number): void {
-		const lastDeletedKey = start + length - 1;
+	public delete(start: K, length: number): void {
+		const lastDeleteKey = this.offsetKey(start, length - 1);
+		for (const { start: key, length: entryLength, value } of this.getIntersectingEntries(
+			start,
+			length,
+		)) {
+			this.tree.delete(key);
+			const lengthBefore = this.subtractKeys(start, key);
+			if (lengthBefore > 0) {
+				// A portion of this entry comes before the deletion range, so we reinsert that portion.
+				this.tree.set(key, { length: lengthBefore, value });
+			}
+
+			const lastEntryKey = this.offsetKey(key, entryLength - 1);
+			const lengthAfter = this.subtractKeys(lastEntryKey, lastDeleteKey);
+			if (lengthAfter > 0) {
+				// A portion of this entry comes after the deletion range, so we reinsert that portion.
+				this.tree.set(this.offsetKey(lastDeleteKey, 1), { length: lengthAfter, value });
+			}
+		}
+	}
+
+	public clone(): RangeMap<K, V> {
+		const cloned = new RangeMap<K, V>(this.offsetKey, this.subtractKeys);
+		cloned.tree = this.tree.clone();
+		return cloned;
+	}
+
+	/**
+	 * Returns a new map which contains the entries from both input maps.
+	 */
+	public static mergeMaps<K, V>(a: RangeMap<K, V>, b: RangeMap<K, V>): RangeMap<K, V> {
```

Comment:
Action verbs like "merge" suggest that side effects may occur. (This would be less ambiguous if we had readonly versions of the RangeMap types.)
Either way, how about naming this `union` instead?